libbeat/processors/add_id/generator/es_generator.go

```diff
@@ -0,0 +1,121 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package generator
+
+import (
+	"encoding/base64"
+	"math/rand"
+	"sync"
+	"time"
+)
+
+type esTimeBasedUUIDGenerator struct{}
+
+// Singleton instance
+var _esTimeBasedUUIDGenerator IDGenerator = (*esTimeBasedUUIDGenerator)(nil)
+
+// ESTimeBasedUUIDGenerator returns the singleton instance for this generator
+func ESTimeBasedUUIDGenerator() IDGenerator {
+	return _esTimeBasedUUIDGenerator
+}
+
+var (
+	sequenceNumber uint32
+	lastTimestamp  time.Time
+	mac            []byte
+	mu             sync.Mutex
+)
+
+func init() {
+	m, err := getSecureMungedMACAddress()
+	if err != nil {
+		panic(err)
+	}
+	mac = m
+	sequenceNumber = rand.Uint32()
+}
+
+// NextID returns a base64-encoded, randomly-generated, but roughly ordered (over time), unique
+// ID. The algorithm used to generate the ID is the same as used by Elasticsearch.
+// See https://github.com/elastic/elasticsearch/blob/a666fb2266/server/src/main/java/org/elasticsearch/common/TimeBasedUUIDGenerator.java
+func (*esTimeBasedUUIDGenerator) NextID() string {
+	mu.Lock()
+	defer mu.Unlock()
+
+	sequenceNumber++
+
+	// We only use bottom 3 bytes for the sequence number.
+	s := sequenceNumber & 0xffffff
+
+	timestamp := getTimestamp()
+	if s == 0 {
+		// Always force the clock to increment whenever sequence number is 0, in case we have a long time-slip backwards.
+		timestamp.Add(1 * time.Millisecond)
+	}
+	lastTimestamp = timestamp
+
+	t := timestamp.UnixNano() / 1000 // timestamp in ms-since-epoch
+
+	uuidBytes := make([]byte, 15)
```

Comment:
we allocate here and then encode to a string which is another allocation. In order to reduce contention on the mutex better split the encoding and the get sequenceNumber + timestamp into separate functions (more fine grained locking). E.g.

```
func (...) NextID string {
  timestamp, s := nextIDData()
  var uuidBytes [15]byte

  packID(uuidbytes[:], timestamp, s)
  return base64.RawURLEncoding.EncodeToString(uuidBytes[:])
}

func nextIDData() (time.Time, uint32) {
  mu.Lock()
  defer mu.Unlock()
  
  s := sequenceNumber & 0xffffff
  timestamp := getTimestamp()
  ...
  return timestamp, s
}

func packID(buf []byte, ts time.Time, seq uint32) {
   ...
}
```